src/category_theory/limits/shapes/zero.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2019 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.limits.shapes.terminal
+
+/-!
+# Zero morphisms and zero objects
+
+A category "has zero morphisms" if there is a designated "zero morphism" in each morphism space,
+and compositions of zero morphisms with anything give the zero morphism. (Notice this is extra
+structure, not merely a property.)
+
+A category "has a zero object" if it has an object which is both initial and terminal. Having a
+zero object provides zero morphisms, as the unique morphisms factoring through the zero object.
+
+## References
+* https://en.wikipedia.org/wiki/Zero_morphism
+-/
+
+universes v u
+
+open category_theory
+
+namespace category_theory.limits
+
+variables (C : Type u) [𝒞 : category.{v} C]
+include 𝒞
+
+/-- A category "has zero morphisms" if there is a designated "zero morphism" in each morphism space,
+and compositions of zero morphisms with anything give the zero morphism. -/
+class has_zero_morphisms :=
+[has_zero : Π X Y : C, has_zero (X ⟶ Y)]
+(comp_zero' : ∀ {X Y : C} (f : X ⟶ Y) (Z : C), f ≫ (0 : Y ⟶ Z) = (0 : X ⟶ Z) . obviously)
+(zero_comp' : ∀ (X : C) {Y Z : C} (f : Y ⟶ Z), (0 : X ⟶ Y) ≫ f = (0 : X ⟶ Z) . obviously)
+
+attribute [instance] has_zero_morphisms.has_zero
+restate_axiom has_zero_morphisms.comp_zero'
+attribute [simp] has_zero_morphisms.comp_zero
+restate_axiom has_zero_morphisms.zero_comp'
+attribute [simp, reassoc] has_zero_morphisms.zero_comp
+
+/-- A category "has a zero object" if it has an object which is both initial and terminal. -/
+class has_zero_object :=
+(zero : C)
+(unique_to : ∀ X : C, unique (zero ⟶ X))
+(unique_from : ∀ X : C, unique (X ⟶ zero))
+
+variables [has_zero_object.{v} C]
+
+variables {C}
+
+/-- We allow writing `0` to indicate the zero object in any category with a zero object. -/
+instance [has_zero_object.{v} C] : has_zero C :=
+{ zero := has_zero_object.zero.{v} C }
+
+attribute [instance] has_zero_object.unique_to has_zero_object.unique_from
+
+namespace has_zero_object
+-- We now make some preliminary definitions in preparation for constructing an instance of
+-- `has_zero_morphisms C`. They can all be private, as public versions have provided by the
+-- `has_zero_morphisms` typeclass.
+
+private def zero_morphism (X Y : C) :=
+inhabited.default (X ⟶ 0) ≫ inhabited.default (0 ⟶ Y)
```

Comment:
I wonder whether `inhabited.default (X ⟶ 0)` is really the way we want to write the unique morphism from X to a zero (or terminal) object.

Reply:
I don't think so. This is a temporary definition. As soon as `has_zero_morphisms` kicks in, there is also `has_terminal` and that comes with it's own way of writing this (I forgot the exact name).